ansible/playbooks/roles/postgres_exporter/tasks/upgrade/main.yml

```diff
@@ -0,0 +1,105 @@
+---
+- name: Check for upgrade flag file
+  stat:
+    path: "{{ lock_file }}"
+    get_attributes: false
+    get_checksum: false
+    get_mime: false
+  register: lock_file_status
+
+- name: Preflight check
+  when: not lock_file_status.stat.exists
+  block:
+  - name: Check for binary existance
+    stat:
+      path: "{{ exporter.binary_path }}"
+      get_attributes: false
+      get_checksum: false
+      get_mime: false
+    register: postgres_exporter_binary
+
+  - name: Inform and fail if postgres-exporter binary is missing
+    when: not postgres_exporter_binary.stat.exists
+    fail:
+      msg: >
+        There is no postgres-exporter binary at all,
+        re-apply your configuration or remove postgres_exporter role from inventory
+
+  - name: Verify installed version
+    shell: /opt/postgres_exporter/postgres_exporter --version 2>&1 | grep postgres_exporter
+    register: postgres_exporter_current_version
+
+  - name: Set version facts
+    set_fact:
+      postgres_exporter_version:
+        old: "{{ postgres_exporter_current_version.stdout_lines[0] | regex_search('\\d+\\.\\d+\\.\\d+') }}"
+        new: "{{ exporter.version }}"
+
+  - name: Avoiding risk of downgrade
+    when: postgres_exporter_version.old is not version( postgres_exporter_version.new, '<' )
+    debug:
+      msg: "Skipping upgrade: Postgres-exporter in the same or newer version already installed!"
+
+- name: Postgres-exporter | Run upgrade tasks
+  when: >
+    postgres_exporter_version.old is version( postgres_exporter_version.new, '<' )
```

Comment:
@cicharka 
`postgres_exporter_version.old/new` variables are undefined when upgrade flag file exists. This will fail the build as the whole block where we set facts is not executed.
Important: probably all upgrade playbooks with this approach are affected.
Possibly reordering the conditions and checking the state file first will fix the problem.

Reply:
Ok fixed now!